### ISSUE TITLE
Update for Beat Saber 1.12.1

### ISSUE THREAD

**Issue by opl-** [opl-/beatsaber-http-status#55]
New version of the game has been released.

There's an untested version of this mod for v1.12.1 on the update/1.12.1 branch. Currently any further work and testing is blocked by `BS_Utils.Plugin.LevelData.GameplayCoreSceneSetupData` returning `null`. https://github.com/opl-/beatsaber-http-status/tree/update/1.12.1

**Comment by alexharl** [comment]
Hey there,
any updates on when you plan on releasing for 1.12?
Thanks and keep up the good work!

**Comment by opl-** [comment]
The 50af3961caa0d5e73d269b2e60a269d37d2376e2 commit technically already works with multiplayer, but the code is messy and then broke while I was trying to clean it up. Unfortunately I got ill while working on that and then haven't been in the right mind state to do any more work, so it kinda stagnated.

I was planning to expose the data from other players in multiplayer before releasing a new version, but since that will require a bit more figuring stuff out, I might instead just release the code basically as is and worry about that later. We'll see.

**Comment by Gordin** [comment]
Hi, I've been trying to compile the 1.12.1 branch (I only care about it working in solo mode and nothing else...), but I can't get past this error:
```
1>------ Build started: Project: BeatSaberHTTPStatusPlugin, Configuration: Debug Any CPU ------
1>C:\Users\Gordin\code\beatsaber-http-status\BeatSaberHTTPStatus\obj\Debug\src\Plugin.cs(253,31,253,39): error CS0117: 'Gamemode' does not contain a definition for 'GameMode'
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```
I think I have the project configured correctly otherwise (Visual Studio shows no missing references). I didn't change any code, so maybe I have the wrong version of one of the referenced Plugin DLLs? Right now I have everything updated through ModAssistant.

**Comment by alexharl** [comment]
@Gordin i also only care about solo/party mode, so maybe @opl- is willing to first release a non-mp version?
Meanwhile, i'll try to build it myself and get back to you

**Comment by opl-** [comment]
Seems Kyle removed the `Gamemode.GameMode` in favor of just accessing data on `LevelData` in this commit https://github.com/Kylemc1413/Beat-Saber-Utils/commit/a2ccce30ba453c16cb7a1065979a9767d4e18491 which, as far as I can tell, was released in v1.6.0. I was testing on an early, unreleased version of the plugin that hasn't yet changed that, hence the errors.

The property can now be accessed using this fun, easy to remember snippet: `BS_Utils.Plugin.LevelData.GameplayCoreSceneSetupData.difficultyBeatmap.parentDifficultyBeatmapSet.beatmapCharacteristic.serializedName`.

I pushed this change to the branch in 7d96af9. Below is a development build that I only tested in standard solo and in multiplayer (I believe spectating hasn't yet been tested properly). No guarantees that things won't break. If they do, feel free to report back here.

[BeatSaberHTTPStatus-1.11.3-dev-bs1.12.1-7d96af9c.zip](https://github.com/opl-/beatsaber-http-status/files/5491215/BeatSaberHTTPStatus-1.11.3-dev-bs1.12.1-7d96af9c.zip)

**Comment by flosoft** [comment]
Just wnated to share that the version posted above also works on 1.13.0

**Comment by tehbiga** [comment]
Also want to add to this that I just had to manually fix the DLL references since my user config failed to take hold and ado a small fix to HTTPServer to get this working for 1.13:

```cs
server.AddWebSocketService<StatusBroadcastBehavior>("/socket", behavior => behavior.SetStatusManager(statusManager));
```
to
```cs
server.AddWebSocketService<StatusBroadcastBehavior>("/socket", () => new StatusBroadcastBehavior(statusManager));
```
_(And also adding the constructor of course)_

**Comment by rynan4818** [comment]
I found two problems with the above version.
The environment we checked is 1.13.0.

1) Unable to get noteID when "No Arrows" is selected.
2) The "obstacleEnter" and "obstacleExit" events do not occur.

The problem in 1) was solved by making the following modifications to "Plugin.cs".

```
public static bool NoteDataEquals(NoteData a, NoteData b) {
	return a.time == b.time && a.lineIndex == b.lineIndex && a.noteLineLayer == b.noteLineLayer && a.colorType == b.colorType && a.cutDirection == b.cutDirection && a.duration == b.duration;
}
```

to

```
public bool NoteDataEquals(NoteData a, NoteData b) {
	if (statusManager.gameStatus.modNoArrows) {
		return a.time == b.time && a.lineIndex == b.lineIndex && a.noteLineLayer == b.noteLineLayer && a.colorType == b.colorType && a.duration == b.duration;
	} else {
		return a.time == b.time && a.lineIndex == b.lineIndex && a.noteLineLayer == b.noteLineLayer && a.colorType == b.colorType && a.cutDirection == b.cutDirection && a.duration == b.duration;
    }
}
```

The problem in 2) could not be solved by me.

**Comment by opl-** [comment]
I can finally close this one. You can download [v1.12.0](https://github.com/opl-/beatsaber-http-status/releases/tag/v1.12.0) from the releases page. It will appear on Beat Mods (and in mod managers) when it gets approved.

If you find any bugs, please open a new issue.

---

@tehbiga I'm not sure what you mean. The version you changed it to is the [deprecated one](https://github.com/sta/websocket-sharp/blob/419a042a657363262e017c086984678bb06bb3e6/websocket-sharp/Server/HttpServer.cs#L1213), which I'm specifically not using. Unless the websocket-sharp version on BeatMods is awfully outdated (which I doubt) or you somehow ended up with an outdated version from git, it should work just fine as is.

@rynan4818 Thanks for pointing those issues out. Both should be fixed.
